app/src/main/kotlin/com/exactpro/th2/lwdataprovider/grpc/GrpcDataProviderBackPressure.kt

```diff
@@ -57,50 +58,90 @@ class GrpcDataProviderBackPressure(
         val servCallObs = responseObserver as ServerCallStreamObserver<T>
         val lock = ReentrantLock()
         var future: Future<*>? = null
+        val isCancelled = AtomicBoolean(false)
 
         fun cleanBuffer() {
             while (buffer.poll() != null) {
                 buffer.clear()
             }
         }
+
         fun cancel() {
-            handler.cancel()
-            onClose(handler)
-            cleanBuffer()
-            onFinished()
+            if (isCancelled.compareAndSet(false, true)) {
+                handler.cancel()
+                onClose(handler)
+                cleanBuffer()
+                onFinished()
+                logger.info { "Stream cancelled and cleaned up" }
+            }
         }
+
+        servCallObs.setOnCancelHandler {
+            logger.warn { "Execution cancelled" }
+            lock.withLock {
+                future?.cancel(true)
+                future = null
+            }
+            cancel()
+        }
+
         servCallObs.setOnReadyHandler {
-            if (!handler.isAlive)
+            if (!handler.isAlive || isCancelled.get()) {
```

Comment:
question: Why is `handler.isAlive` not enough here? The `isAlive` should return `false` in the same case when `isCanceled` is true